arnglld/src/main.rs

```diff
@@ -33,19 +40,175 @@ struct Opt {
     #[clap(short, long, parse(from_occurrences))]
     verbose: usize,
 
+    /// List audio devices
+    #[clap(long)]
+    list_devices: bool,
+
     #[clap(short, long)]
-    callsign: HamAddr,
+    callsign: Option<HamAddr>,
+
+    #[clap(long)]
+    input_audio_device: Option<String>,
+
+    #[clap(long)]
+    output_audio_device: Option<String>,
+}
+
+impl Opt {
+    fn get_output_device(&self) -> Result<cpal::Device, anyhow::Error> {
+        let host = cpal::default_host();
+
+        if let Some(name) = self.output_audio_device.as_ref() {
+            let lc_name = name.to_lowercase();
+
+            // First try exact matches
+            for device in host.output_devices()? {
+                let dev_name = device.name()?.to_lowercase();
+                if dev_name == lc_name {
+                    return Ok(device);
+                }
+            }
+
+            // Then try prefixes
+            for device in host.output_devices()? {
+                let dev_name = device.name()?.to_lowercase();
+                if dev_name.starts_with(&lc_name) {
+                    return Ok(device);
+                }
+            }
+
+            // Then try for any device that matches the substring
+            for device in host.output_devices()? {
+                let dev_name = device.name()?.to_lowercase();
+                if dev_name.contains(&lc_name) {
+                    return Ok(device);
+                }
+            }
+
+            return Err(format_err!("Cannot find output device matching {:?}", name));
+        }
+
+        host.default_output_device()
+            .ok_or_else(|| format_err!("no default output device"))
+    }
+
+    #[allow(dead_code)]
+    fn get_input_device(&self) -> Result<cpal::Device, anyhow::Error> {
+        let host = cpal::default_host();
+
+        if let Some(name) = self.input_audio_device.as_ref() {
```

Comment:
Generally the same here. We could definitely extract the logic of finding a named device into a helper.

Reply:
Yes, I'll go ahead and add a helper.